File: backend/internal/models/follow.go
package models

import (
	"time"
	"gorm.io/gorm"
)

// FollowStatus 关注状态枚举
type FollowStatus string

const (
	FollowStatusActive  FollowStatus = "active"  // 正常关注
	FollowStatusBlocked FollowStatus = "blocked" // 被屏蔽
	FollowStatusMuted   FollowStatus = "muted"   // 静音
)

// UserRelationship 用户关系模型
type UserRelationship struct {
	ID                  string       `json:"id" gorm:"primaryKey;type:varchar(36)"`
	FollowerID          string       `json:"follower_id" gorm:"type:varchar(36);index"`
	FollowingID         string       `json:"following_id" gorm:"type:varchar(36);index"`
	Status              FollowStatus `json:"status" gorm:"type:varchar(20);default:'active'"`
	NotificationEnabled bool         `json:"notification_enabled" gorm:"default:true"`
	CreatedAt           time.Time    `json:"created_at"`
	UpdatedAt           time.Time    `json:"updated_at"`

	// 关联
	Follower  User `json:"follower,omitempty" gorm:"foreignKey:FollowerID"`
	Following User `json:"following,omitempty" gorm:"foreignKey:FollowingID"`
}

// FollowUser 关注用户响应结构（匹配前端类型）
type FollowUser struct {
	ID                    string    `json:"id"`
	Username              string    `json:"username"`
	Nickname              string    `json:"nickname"`
	Avatar                string    `json:"avatar"`
	Bio                   string    `json:"bio"`
	School                string    `json:"school"`
	OPCode                string    `json:"op_code"`
	WritingLevel          int       `json:"writing_level"`
	CourierLevel          int       `json:"courier_level"`
	FollowersCount        int       `json:"followers_count"`
	FollowingCount        int       `json:"following_count"`
	LettersCount          int       `json:"letters_count"`
	IsFollowing           bool      `json:"is_following,omitempty"`
	IsFollower            bool      `json:"is_follower,omitempty"`
	FollowStatus          string    `json:"follow_status,omitempty"`
	FollowedAt            string    `json:"followed_at,omitempty"`
	MutualFollowersCount  int       `json:"mutual_followers_count,omitempty"`
}

// FollowStats 关注统计
type FollowStats struct {
	UserID              string `json:"user_id" gorm:"primaryKey;type:varchar(36)"`
	FollowersCount      int    `json:"followers_count" gorm:"default:0"`
	FollowingCount      int    `json:"following_count" gorm:"default:0"`
	MutualFollowsCount  int    `json:"mutual_follows_count" gorm:"default:0"`
	UpdatedAt           time.Time `json:"updated_at"`
}

// FollowActivity 关注活动记录
type FollowActivity struct {
	ID        string    `json:"id" gorm:"primaryKey;type:varchar(36)"`
	ActorID   string    `json:"actor_id" gorm:"type:varchar(36);index"`
	TargetID  string    `json:"target_id" gorm:"type:varchar(36);index"`
	Type      string    `json:"type" gorm:"type:varchar(50)"`
	CreatedAt time.Time `json:"created_at"`

	// 关联
	Actor  User `json:"actor,omitempty" gorm:"foreignKey:ActorID"`
	Target User `json:"target,omitempty" gorm:"foreignKey:TargetID"`
}

// UserSuggestion 用户推荐
type UserSuggestion struct {
	ID              string    `json:"id" gorm:"primaryKey;type:varchar(36)"`
	UserID          string    `json:"user_id" gorm:"type:varchar(36);index"`
	SuggestedUserID string    `json:"suggested_user_id" gorm:"type:varchar(36);index"`
	Reason          string    `json:"reason" gorm:"type:varchar(100)"`
	Score           float64   `json:"score" gorm:"type:decimal(5,4)"`
	CreatedAt       time.Time `json:"created_at"`
	
	// 关联
	User          User `json:"user,omitempty" gorm:"foreignKey:UserID"`
	SuggestedUser User `json:"suggested_user,omitempty" gorm:"foreignKey:SuggestedUserID"`
}

// Request/Response 结构

// FollowActionRequest 关注操作请求
type FollowActionRequest struct {
	UserID              string `json:"user_id" binding:"required"`
	NotificationEnabled bool   `json:"notification_enabled"`
}

// FollowActionResponse 关注操作响应
type FollowActionResponse struct {
	Success        bool   `json:"success"`
	IsFollowing    bool   `json:"is_following"`
	FollowerCount  int    `json:"follower_count"`
	FollowingCount int    `json:"following_count"`
	FollowedAt     string `json:"followed_at,omitempty"`
	Message        string `json:"message,omitempty"`
}

// FollowListRequest 关注列表请求
type FollowListRequest struct {
	Page         int      `json:"page" form:"page"`
	Limit        int      `json:"limit" form:"limit"`
	SortBy       string   `json:"sort_by" form:"sort_by"`
	Order        string   `json:"order" form:"order"`
	Search       string   `json:"search" form:"search"`
	SchoolFilter string   `json:"school_filter" form:"school_filter"`
	StatusFilter []string `json:"status_filter" form:"status_filter"`
}

// FollowListResponse 关注列表响应
type FollowListResponse struct {
	Users []FollowUser `json:"users"`
	Pagination struct {
		Page  int `json:"page"`
		Limit int `json:"limit"`
		Total int `json:"total"`
		Pages int `json:"pages"`
	} `json:"pagination"`
}

// UserSearchRequest 用户搜索请求
type UserSearchRequest struct {
	Query        string `json:"query" form:"query"`
	SchoolCode   string `json:"school_code" form:"school_code"`
	Role         string `json:"role" form:"role"`
	MinFollowers int    `json:"min_followers" form:"min_followers"`
	MaxFollowers int    `json:"max_followers" form:"max_followers"`
	ActiveSince  string `json:"active_since" form:"active_since"`
	SortBy       string `json:"sort_by" form:"sort_by"`
	Order        string `json:"order" form:"order"`
	Limit        int    `json:"limit" form:"limit"`
	Offset       int    `json:"offset" form:"offset"`
}

// UserSearchResponse 用户搜索响应
type UserSearchResponse struct {
	Users           []FollowUser `json:"users"`
	Total           int          `json:"total"`
	Query           string       `json:"query"`
	Suggestions     []string     `json:"suggestions,omitempty"`
	FiltersApplied  interface{}  `json:"filters_applied"`
}

// UserSuggestionsRequest 用户推荐请求
type UserSuggestionsRequest struct {
	Limit             int     `json:"limit" form:"limit"`
	BasedOn           string  `json:"based_on" form:"based_on"`
	ExcludeFollowed   bool    `json:"exclude_followed" form:"exclude_followed"`
	MinActivityScore  float64 `json:"min_activity_score" form:"min_activity_score"`
}

// UserSuggestionsResponse 用户推荐响应
type UserSuggestionsResponse struct {
	Suggestions         []FollowSuggestionItem `json:"suggestions"`
	AlgorithmUsed       string                 `json:"algorithm_used"`
	RefreshAvailableAt  string                 `json:"refresh_available_at"`
}

// FollowSuggestionItem 推荐项
type FollowSuggestionItem struct {
	User             FollowUser   `json:"user"`
	Reason           string       `json:"reason"`
	ConfidenceScore  float64      `json:"confidence_score"`
	MutualFollowers  []FollowUser `json:"mutual_followers,omitempty"`
	CommonInterests  []string     `json:"common_interests,omitempty"`
}

// FollowStatsResponse 关注统计响应
type FollowStatsResponse struct {
	FollowersCount       int          `json:"followers_count"`
	FollowingCount       int          `json:"following_count"`
	MutualFollowersCount int          `json:"mutual_followers_count"`
	RecentFollowers      []FollowUser `json:"recent_followers"`
	PopularFollowing     []FollowUser `json:"popular_following"`
}

// TableName 指定表名
func (UserRelationship) TableName() string {
	return "user_relationships"
}

func (FollowStats) TableName() string {
	return "user_follow_stats"
}

func (FollowActivity) TableName() string {
	return "follow_activities"
}

func (UserSuggestion) TableName() string {
	return "user_suggestions"
}

// BeforeCreate GORM钩子 - 生成UUID
func (ur *UserRelationship) BeforeCreate(tx *gorm.DB) error {
	if ur.ID == "" {
		ur.ID = generateUUID()
	}
	return nil
}

func (fa *FollowActivity) BeforeCreate(tx *gorm.DB) error {
	if fa.ID == "" {
		fa.ID = generateUUID()
	}
	return nil
}

func (us *UserSuggestion) BeforeCreate(tx *gorm.DB) error {
	if us.ID == "" {
		us.ID = generateUUID()
	}
	return nil
}

// IsActive 检查关注关系是否为活跃状态
func (ur *UserRelationship) IsActive() bool {
	return ur.Status == FollowStatusActive
}

// CanReceiveNotifications 检查是否可以接收通知
func (ur *UserRelationship) CanReceiveNotifications() bool {
	return ur.IsActive() && ur.NotificationEnabled
}

// ToFollowUser 将User转换为FollowUser
func (u *User) ToFollowUser(stats *FollowStats, profile *UserProfileExtended) *FollowUser {
	followUser := &FollowUser{
		ID:           u.ID,
		Username:     u.Username,
		Nickname:     u.Nickname,
		Avatar:       u.Avatar,
		WritingLevel: 1, // Default
		CourierLevel: 0, // Default
	}

	// 添加统计信息
	if stats != nil {
		followUser.FollowersCount = stats.FollowersCount
		followUser.FollowingCount = stats.FollowingCount
	}

	// 添加档案信息
	if profile != nil {
		followUser.Bio = profile.Bio
		followUser.School = profile.School
		followUser.OPCode = profile.OPCode
		followUser.WritingLevel = profile.WritingLevel
		followUser.CourierLevel = profile.CourierLevel
	}

	return followUser
}

// Note: generateUUID and generateRandomString functions are already defined in other model files

File: frontend/src/types/follow.ts
/**
 * Follow System Types - SOTA Implementation
 * 关注/粉丝系统类型定义 - 支持用户关系管理、社交互动、推荐系统
 */

import type { User } from './user'

// ================================
// Core Follow Relationship Types
// ================================

export interface FollowRelationship {
  id: string
  follower_id: string
  following_id: string
  created_at: string
  updated_at: string
  status: FollowStatus
  notification_enabled: boolean
}

export type FollowStatus = 'active' | 'blocked' | 'muted'

export interface FollowUser extends User {
  // Extended user info for follow contexts
  followers_count: number
  following_count: number
  letters_count: number
  is_following?: boolean
  is_follower?: boolean
  follow_status?: FollowStatus
  followed_at?: string
  mutual_followers_count?: number
}

// ================================
// API Request/Response Types
// ================================

export interface FollowActionRequest {
  user_id: string
  notification_enabled?: boolean
}

export interface FollowActionResponse {
  success: boolean
  is_following: boolean
  follower_count: number
  following_count: number
  followed_at?: string
  message?: string
}

export interface FollowListQuery {
  page?: number
  limit?: number
  sort_by?: 'created_at' | 'nickname' | 'letters_count'
  order?: 'asc' | 'desc'
  search?: string
  school_filter?: string
  status_filter?: FollowStatus[]
}

export interface FollowListResponse {
  users: FollowUser[]
  pagination: {
    page: number
    limit: number
    total: number
    pages: number
  }
}

export interface FollowStatsResponse {
  followers_count: number
  following_count: number
  mutual_followers_count: number
  recent_followers: FollowUser[]
  popular_following: FollowUser[]
}

export interface UserSuggestionsQuery {
  limit?: number
  based_on?: 'school' | 'mutual_followers' | 'activity' | 'interests'
  exclude_followed?: boolean
  min_activity_score?: number
}

export interface UserSuggestionsResponse {
  suggestions: FollowSuggestion[]
  algorithm_used: string
  refresh_available_at: string
}

export interface UserSearchQuery {
  query: string
  page?: number
  limit?: number
  sort_by?: 'activity' | 'followers' | 'joined' | 'relevance'
  order?: 'asc' | 'desc'
  search?: string
  school_filter?: string
  status_filter?: FollowStatus[]
}

export interface FollowSuggestion {
  user: FollowUser
  reason: SuggestionReason
  confidence_score: number
  mutual_followers?: FollowUser[]
  common_interests?: string[]
}

export type SuggestionReason = 
  | 'same_school'
  | 'mutual_followers'
  | 'similar_interests'
  | 'active_user'
  | 'new_user'
  | 'trending'

// ================================
// Follow Activity & Notifications
// ================================

export interface FollowActivity {
  id: string
  type: FollowActivityType
  actor_id: string
  target_id: string
  actor: FollowUser
  target?: FollowUser
  created_at: string
  is_read: boolean
  metadata?: Record<string, any>
}

export type FollowActivityType = 
  | 'new_follower'
  | 'followed_you_back'
  | 'user_joined'
  | 'milestone_reached'

export interface FollowNotificationSettings {
  new_followers: boolean
  follow_backs: boolean
  suggestions: boolean
  activity_digest: boolean
  email_notifications: boolean
}

// ================================
// UI Component Props
// ================================

export interface FollowButtonProps {
  user_id: string
  initial_is_following?: boolean
  initial_follower_count?: number
  size?: 'sm' | 'md' | 'lg'
  variant?: 'default' | 'outline' | 'ghost'
  show_count?: boolean
  disabled?: boolean
  className?: string
  onFollowChange?: (isFollowing: boolean, followerCount: number) => void
}

export interface FollowListProps {
  user_id?: string // If not provided, use current user
  type: 'followers' | 'following'
  initial_data?: FollowUser[]
  show_stats?: boolean
  enable_search?: boolean
  enable_filters?: boolean
  max_height?: string
  className?: string
}

export interface FollowStatsProps {
  user_id?: string
  show_detailed?: boolean
  show_recent?: boolean
  compact?: boolean
  className?: string
}

export interface UserSuggestionsProps {
  limit?: number
  show_reason?: boolean
  show_mutual?: boolean
  show_refresh?: boolean
  algorithm?: UserSuggestionsQuery['based_on']
  className?: string
  onUserFollow?: (user: FollowUser) => void
}

export interface FollowActivityFeedProps {
  user_id?: string
  limit?: number
  show_own_actions?: boolean
  show_avatars?: boolean
  real_time?: boolean
  className?: string
}

// ================================
// Store State Types
// ================================

export interface FollowState {
  // Current user's follow data
  followers: FollowUser[]
  following: FollowUser[]
  follower_count: number
  following_count: number
  
  // Cached data
  suggestions: FollowSuggestion[]
  recent_activities: FollowActivity[]
  
  // UI state
  loading: {
    followers: boolean
    following: boolean
    suggestions: boolean
    follow_action: boolean
  }
  
  // Last updated timestamps
  last_updated: {
    followers: number | null
    following: number | null
    suggestions: number | null
  }
  
  // Error states
  errors: {
    followers: string | null
    following: string | null
    suggestions: string | null
    follow_action: string | null
  }
  
  // Settings
  notification_settings: FollowNotificationSettings
}

export interface FollowActions {
  // Core actions
  followUser: (userId: string, options?: { notificationEnabled?: boolean }) => Promise<boolean>
  unfollowUser: (userId: string) => Promise<boolean>
  
  // Data fetching
  loadFollowers: (userId?: string, query?: FollowListQuery) => Promise<void>
  loadFollowing: (userId?: string, query?: FollowListQuery) => Promise<void>
  loadSuggestions: (query?: UserSuggestionsQuery) => Promise<void>
  refreshSuggestions: () => Promise<void>
  
  // Bulk actions
  followMultipleUsers: (userIds: string[]) => Promise<{ success: string[]; failed: string[] }>
  removeFollower: (userId: string) => Promise<boolean>
  
  // User data
  getUserFollowStatus: (userId: string) => Promise<{ isFollowing: boolean; isFollower: boolean }>
  searchUsers: (query: string, filters?: Partial<FollowListQuery>) => Promise<FollowUser[]>
  
  // Settings
  updateNotificationSettings: (settings: Partial<FollowNotificationSettings>) => Promise<void>
  
  // Cache management
  clearCache: () => void
  invalidateUser: (userId: string) => void
}

// ================================
// Utility Types
// ================================

export interface FollowRelationshipSummary {
  mutual_followers: FollowUser[]
  followers_only: FollowUser[]
  following_only: FollowUser[]
  suggestions_based_on_mutual: FollowSuggestion[]
}

export interface FollowMetrics {
  growth_rate: {
    followers_weekly: number
    following_weekly: number
  }
  engagement_rate: number
  mutual_follow_rate: number
  suggestion_acceptance_rate: number
}

export interface FollowError {
  code?: string
  message: string
  status?: number
  details?: Record<string, any>
}

// ================================
// Search and Discovery
// ================================

export interface UserSearchQuery {
  query?: string
  school_code?: string
  role?: string
  min_followers?: number
  max_followers?: number
  active_since?: string
  sort_by?: 'followers' | 'activity' | 'joined' | 'relevance'
  order?: 'asc' | 'desc'
  limit?: number
  offset?: number
}

export interface UserSearchResponse {
  users: FollowUser[]
  total: number
  query: string
  suggestions?: string[]
  filters_applied: Partial<UserSearchQuery>
}

// ================================
// Real-time Updates
// ================================

export interface FollowWebSocketEvent {
  type: 'follow_action' | 'follow_stats_update' | 'new_suggestion'
  data: {
    user_id: string
    target_id?: string
    action?: 'follow' | 'unfollow'
    new_stats?: Partial<FollowStatsResponse>
    suggestion?: FollowSuggestion
  }
  timestamp: string
}

// ================================
// Export convenience types
// ================================

export type {
  User as BaseUser,
}

// Default exports for common interfaces
export type FollowSystemUser = FollowUser
export type FollowSystemState = FollowState & FollowActions